File: android/vault-access/src/test/kotlin/org/secretary/browse/FakeTrashPort.kt
package org.secretary.browse

import kotlinx.coroutines.CompletableDeferred

/**
 * In-memory [TrashPort] test double. Seed [list]/[expired]/reports; prime [listError] to make
 * [listTrashedBlocks] throw. Records each write for assertions. Mirror of iOS `FakeTrashPort`.
 *
 * When [writeGate] is set, every suspend write awaits it before recording the call — lets a
 * test hold a write in flight while a second call hits the model's re-entrancy guard (a
 * faithful race, not a sleep).
 */
class FakeTrashPort(
    var list: List<TrashedBlockInfo> = emptyList(),
    var expired: List<ExpiredEntryInfo> = emptyList(),
    private val listError: VaultBrowseError? = null,
    private val windowMs: Long = 90L * MS_PER_DAY,
    private val writeGate: CompletableDeferred<Unit>? = null,
    var emptyTrashFilesFailed: Int = 0,
    var retentionFilesFailed: Int = 0,
) : TrashPort {
    val restored = mutableListOf<ByteArray>()
    val purged = mutableListOf<ByteArray>()
    var emptied = 0
    val autoPurged = mutableListOf<Long>()

    override fun listTrashedBlocks(): List<TrashedBlockInfo> {
        listError?.let { throw it }
        return list
    }

    override fun expiredTrashEntries(windowMs: Long): List<ExpiredEntryInfo> = expired

    override fun defaultRetentionWindowMs(): Long = windowMs

    override suspend fun restoreBlock(uuid: ByteArray) {
        writeGate?.await()
        restored += uuid
    }

    override suspend fun purgeBlock(uuid: ByteArray): PurgeResultInfo {
        writeGate?.await()
        purged += uuid
        return PurgeResultInfo(uuid, wasShared = false, recipientCount = 0, filesRemoved = 1)
    }

    override suspend fun emptyTrash(): EmptyTrashReportInfo {
        writeGate?.await()
        emptied += 1
        return EmptyTrashReportInfo(list.size, 0, list.size, 0, list.size, emptyTrashFilesFailed)
    }

    override suspend fun autoPurgeExpired(windowMs: Long): RetentionReportInfo {
        writeGate?.await()
        autoPurged += windowMs
        return RetentionReportInfo(expired.size, 0, expired.size, 0, expired.size, retentionFilesFailed, windowMs)
    }
}

File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/FakeTrashPort.swift
import SecretaryVaultAccess

/// In-memory `TrashPort` double with spies + write-failure injection,
/// modeled on `FakeVaultSession`.
public final class FakeTrashPort: TrashPort, @unchecked Sendable {
    public var trashedBlocks: [TrashedBlockInfo]
    public var expiredEntries: [ExpiredEntryInfo]
    public var defaultWindowMs: UInt64
    public var emptyTrashFilesFailed: UInt32 = 0
    public var failNextWrite: VaultAccessError?

    public private(set) var listCount = 0
    public private(set) var previewCount = 0
    public private(set) var restoredUuids: [[UInt8]] = []
    public private(set) var purgedUuids: [[UInt8]] = []
    public private(set) var emptyTrashCount = 0
    public private(set) var autoPurgeWindows: [UInt64] = []

    public init(trashedBlocks: [TrashedBlockInfo] = [],
                expiredEntries: [ExpiredEntryInfo] = [],
                defaultWindowMs: UInt64 = 90 * 86_400_000) {
        self.trashedBlocks = trashedBlocks
        self.expiredEntries = expiredEntries
        self.defaultWindowMs = defaultWindowMs
    }

    private func throwIfInjected() throws {
        if let e = failNextWrite { failNextWrite = nil; throw e }
    }

    public func listTrashedBlocks() throws -> [TrashedBlockInfo] {
        listCount += 1
        return trashedBlocks
    }

    public func expiredTrashEntries(windowMs: UInt64) -> [ExpiredEntryInfo] {
        previewCount += 1
        return expiredEntries
    }

    public func defaultRetentionWindowMs() -> UInt64 { defaultWindowMs }

    public func restoreBlock(uuid: [UInt8]) throws {
        try throwIfInjected()
        restoredUuids.append(uuid)
        trashedBlocks.removeAll { $0.blockUuid == uuid }
    }

    public func purgeBlock(uuid: [UInt8]) throws -> PurgeResultInfo {
        try throwIfInjected()
        purgedUuids.append(uuid)
        trashedBlocks.removeAll { $0.blockUuid == uuid }
        return PurgeResultInfo(blockUuid: uuid, wasShared: false,
                               recipientCount: 0, filesRemoved: 1)
    }

    public func emptyTrash() throws -> EmptyTrashReportInfo {
        try throwIfInjected()
        emptyTrashCount += 1
        let n = UInt32(trashedBlocks.count)
        trashedBlocks.removeAll()
        return EmptyTrashReportInfo(purgedCount: n, sharedCount: 0, ownerOnlyCount: n,
                                    unknownCount: 0, filesRemoved: n, filesFailed: emptyTrashFilesFailed)
    }

    public func autoPurgeExpired(windowMs: UInt64) throws -> RetentionReportInfo {
        try throwIfInjected()
        autoPurgeWindows.append(windowMs)
        let n = UInt32(expiredEntries.count)
        return RetentionReportInfo(purgedCount: n, sharedCount: 0, ownerOnlyCount: n,
                                   unknownCount: 0, filesRemoved: n, filesFailed: 0,
                                   windowMs: windowMs)
    }
}
